test/e2e/framework/nodes.go

```diff
@@ -68,3 +69,37 @@ func (f *Framework) SetGatewayLabelOnNode(cluster ClusterIndex, nodeName string,
 			return err
 		})
 }
+
+// FindAnyNonGatewayRouteAgentPodNodes looks for the route agent pod on any cluster node, including control plane
+// The function will be used in upstream ci environment, lack of resources where two GW nodes are used.
+func (f *Framework) FindAnyNonGatewayRouteAgentPodNodes(cluster ClusterIndex) []*v1.Node {
+	gwNodesSet := stringset.New()
+	gwNodes := f.FindNodesByGatewayLabel(cluster, true)
+
+	for _, node := range gwNodes {
+		gwNodesSet.Add(node.Name)
+	}
+
+	var selectedNode string
+	routeAgentPods := f.AwaitPodsByAppLabel(cluster, RouteAgent, TestContext.SubmarinerNamespace, -1)
+
+	for i := range routeAgentPods.Items {
+		if !gwNodesSet.Contains(routeAgentPods.Items[i].Spec.NodeName) {
+			selectedNode = routeAgentPods.Items[i].Spec.NodeName
+		}
+	}
+
+	searchNodes := AwaitUntil("list nodes", func() (interface{}, error) {
+		// Ignore the control plane node labeled as master as it doesn't allow scheduling of pods
+		return KubeClients[cluster].CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{
+			LabelSelector: "kubernetes.io/hostname=" + selectedNode,
+		})
+	}, NoopCheckResult).(*v1.NodeList)
+
+	nodes := []*v1.Node{}
+	for i := range searchNodes.Items {
+		nodes = append(nodes, &searchNodes.Items[i])
+	}
+
+	return nodes
```

Comment:
You can just return `searchNodes` here to simplify, ie instead of returning `[]*v1.Node`.

```suggestion
	return AwaitUntil("list nodes", func() (interface{}, error) {
		// Ignore the control plane node labeled as master as it doesn't allow scheduling of pods
		return KubeClients[cluster].CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{
			LabelSelector: "kubernetes.io/hostname=" + selectedNode,
		})
	}, NoopCheckResult).(*v1.NodeList)
```

Reply:
ack

Reply:
Hi @tpantelis 

Regarding your suggestion to return `*v1.NodeList` instead of `[]*v1.Node`.

The reason I'm returning `[]*v1.Node` data structure is because in within the RouteAgentRestart test, I'm doing the following flow.
```
	nodes := f.FindNodesByGatewayLabel(framework.ClusterA, onGateway)
	// This will happen only in "kind" environment because of the flow of
	// testing gateway failover test
	if len(nodes) == 0 && !onGateway {
		By(fmt.Sprintf("No Non Gateway worker nodes found in %q. Looking for non worker nodes", clusterAName))
		nodes = f.FindAnyNonGatewayRouteAgentPodNodes(framework.ClusterA)
	}
```

And since I'm rewriting the `nodes` variable and `f.FindNodesByGatewayLabel` function returns `[]*v1.Node` data structure, I need to return the same structure.

What do you think, please?

Reply:
OK - that would make it consistent.